### PR TITLE
Fix possible ip spoofing while searching for upnp gateway

### PR DESCRIPTION
**Pull request by tomaszklak**
In addition, direct dependencies are shrank and updated, and sync and async implementations are refactored to have less duplication between them.

Commit by commit review might be easier.

### CODE REVIEW

File: src/aio/search.rs [hunk 1]
```diff
@@ -200,4 +235,130 @@ mod tests {
...
+        <action>
+        </action>
+    </actionList>
+    </root>
+    "#;
+
+    async fn start_http_server(responses: Vec<String>) -> u16 {
+        let addr = SocketAddr::from(([0, 0, 0, 0], 0));
+
+        // We create a TcpListener and bind it to 127.0.0.1:3000
```
Comment (LukasPukenis) [inline src/aio/search.rs]: where does the 3000 come from?

File: src/common/tests.rs [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+            .await
+            .unwrap()
+    })
+    .await;
+    #[cfg(feature = "aio")]
+    aux(|opt| async { crate::aio::search_gateway(opt).await.map(|_| ()) }).await;
+}
+
+#[test(tokio::test)]
+async fn ip_spoofing_in_getxml_body() {
```
Comment (LukasPukenis) [inline src/common/tests.rs]: Is there any test that checks for no spoofing? this file presents spoofing, would be nice to have the opposite as well(if I missed them)

**Review by LukasPukenis — COMMENTED** [review]
+1 The testcases are nice. Left comment about adding a testcase for non-spoofed scenario

File: src/aio/search.rs (hunk 1 above)
Comment (tomaszklak) [reply src/aio/search.rs]: That's a mistake - copy pasted skeleton code from hyper server guide, but forgot to update the comment after I changed the code to all OS to select interface/port.

Fixed.

File: src/common/tests.rs (hunk 2 above)
Comment (tomaszklak) [reply src/common/tests.rs]: Good point, I've added a test for non spoofed case.

File: src/search.rs [hunk 3]
```diff
@@ -29,56 +29,56 @@ pub fn search_gateway(options: SearchOptions) -> Result<Gateway, SearchError> {
...
+    let (control_schema_url, control_url) = match get_control_urls(&addr, &root_url) {
+        Ok(o) => o,
+        Err(e) => {
+            debug!(
+                "Error has occurred while getting control urls. error: {}, addr: {}, root_url: {}",
+                e, addr, root_url
+            );
+            return Err(e);
+        }
+    };
```
Comment (stalowyjez) [inline src/search.rs]: It looks like `map_err` maybe?

Comment (stalowyjez) [inline src/search.rs]: looks a bit like `map_error` maybe?

File: src/common/tests.rs [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+            }
+        }
+    });
+
+    local_port
+}
+
+#[test(tokio::test)]
+async fn ip_spoofing_in_broadcast_response() {
+    async fn aux<F, Fut>(search_gateway: F)
```
Comment (stalowyjez) [inline src/common/tests.rs]: It looks a bit like something which could be wrapped with macro. If not, maybe some comment on this design would be helpful?

**Review by stalowyjez — DISMISSED** [review]
lgtm, +1

**Review by stalowyjez — DISMISSED** [review]

**Review by stalowyjez — APPROVED** [review]

**Pull request merged**